src/vs/workbench/contrib/mcp/browser/mcpCommands.ts

```diff
@@ -0,0 +1,313 @@
+				action: 'stop'
+			});
+			items.push({
+				label: localize2('mcp.restart', 'Restart Server').value,
+				action: 'restart'
+			});
+		}
+
+		items.push({
+			label: localize2('mcp.showOutput', 'Show Output').value,
+			action: 'showOutput'
+		});
+
+		const pick = await quickInputService.pick(items, {
+			placeHolder: localize('mcp.selectAction', 'Select Server Action')
+		});
+
+		if (!pick) {
+			return;
+		}
+
+		switch (pick.action) {
+			case 'start':
+				await server.start();
+				server.showOutput();
+				break;
+			case 'stop':
+				await server.stop();
+				break;
+			case 'restart':
+				await server.stop();
+				await server.start();
+				break;
+			case 'showOutput':
+				server.showOutput();
+				break;
+		}
+	}
+}
+
+
+export class AttachMCPToolsAction extends Action2 {
+
+	static readonly id = 'workbench.action.chat.mcp.attachMcpTools';
+
+	constructor() {
+		super({
+			id: AttachMCPToolsAction.id,
+			title: localize2('workbench.action.chat.editing.attachContext.shortLabel', "Select Tools..."),
+			icon: Codicon.tools,
+			f1: false,
+			category: CHAT_CATEGORY,
+			precondition: ContextKeyExpr.and(
+				McpContextKeys.serverCount.greater(0),
+				ChatContextKeys.location.isEqualTo(ChatAgentLocation.EditingSession)
+			),
+			menu: {
+				when: ContextKeyExpr.and(
+					McpContextKeys.serverCount.greater(0),
+					ChatContextKeys.location.isEqualTo(ChatAgentLocation.EditingSession)
+				),
+				id: MenuId.ChatInputAttachmentToolbar,
+				group: 'navigation'
+			},
+			keybinding: {
+				when: ContextKeyExpr.and(ChatContextKeys.inChatInput, ChatContextKeys.location.isEqualTo(ChatAgentLocation.EditingSession)),
+				primary: KeyMod.CtrlCmd | KeyMod.Shift | KeyCode.Slash,
+				weight: KeybindingWeight.EditorContrib
+			}
+		});
+	}
+
+	override async run(accessor: ServicesAccessor, ...args: any[]): Promise<void> {
+
+		const quickPickService = accessor.get(IQuickInputService);
+		const mcpService = accessor.get(IMcpService);
+
+		type IToolPick = IQuickPickItem & { tool: IMcpTool };
+		const picks: (IToolPick | IQuickPickSeparator)[] = [];
+
+		for (const server of mcpService.servers.get()) {
+
+			picks.push({
+				type: 'separator',
+				label: server.definition.label
+			});
+
+			for (const tool of server.tools.get()) {
+				picks.push({
+					type: 'item',
+					label: tool.definition.name,
+					detail: tool.definition.description,
+					tooltip: tool.definition.description,
+					picked: tool.enabled.get(),
+					tool,
+				});
+			}
+		}
+
+		const result = await quickPickService.pick(picks, {
+			placeHolder: localize('placeholder', "Select tools that are available to chat"),
+			canPickMany: true
+		});
+
+		if (!result) {
+			return;
+		}
+
+		const seen = new Set<IMcpTool>();
+		for (const item of result) {
+			item.tool.updateEnablement(true);
+			seen.add(item.tool);
+		}
+
+		for (const pick of picks) {
+			if (pick.type === 'item' && !seen.has(pick.tool)) {
+				pick.tool.updateEnablement(false);
```

Comment:
I suppose in the future the enablement state could be per-chat view, wouldn't worry about it for now...

Reply:
Yeah, I was thinking about this and wanna do that but ATM we don't have a way to express this. Tools only have the global `when`-clause and doing something session-specific would require new API which enables certain tools only for certain sessions. Like sending along tools with the request like `ChatRequest#toolReferences` but without the reference treatment, e.g without the rendering